subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/LazyConsumerActionExecutor.java

```diff
@@ -81,12 +82,21 @@ public String getDisplayName() {
             BuildCancellationToken cancellationToken = parameters.getCancellationToken();
             InternalBuildProgressListener buildProgressListener = parameters.getBuildProgressListener();
             ConsumerConnection connection = onStartAction(cancellationToken, buildProgressListener);
+            checkProviderVersion(parameters);
+
             return action.run(connection);
         } finally {
             onEndAction();
         }
     }
 
+    private void checkProviderVersion(ConsumerOperationParameters parameters) {
+        if (connection instanceof ParameterValidatingConsumerConnection) {
```

Comment:
❌ The fact that we need to cast and get some internal details tells us that this check doesn't belong here. Instead, it should be a wrapper around the `ConsumerConnection`, created by the tooling implementation loader.